src/NUnitFramework/framework/Constraints/CollectionTally.cs

```diff
@@ -170,10 +170,8 @@ void TryRemoveSlow(IEnumerable c)
 
         private static ArrayList ToArrayList(IEnumerable items)
         {
-            if (items is ICollection ic)
-                return new ArrayList(ic);
+            var list = items is ICollection ic ? new ArrayList(ic.Count) : new ArrayList();
```

Comment:
`ImmutableDictionary.CopyTo` will convert the underlying type which is why the ctor which took in an ICollection was failing, but we can at least initialize the ArrayList to the proper size if it is known